File: kernel/src/arch/pic.c
/*
 * Copyright (c) 2022-present, SkillerRaptor <skillerraptor@protonmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

#include "arch/pic.h"

#include "arch/io.h"

#define MASTER_COMMAND_SELECTOR 0x20
#define MASTER_DATA_SELECTOR 0x21
#define MASTER_OFFSET 0x20

#define SLAVE_COMMAND_SELECTOR 0xa0
#define SLAVE_DATA_SELECTOR 0xa1
#define SLAVE_OFFSET 0x28

#define MASTER_IDENTITY 0x04
#define SLAVE_IDENTITY 0x02

#define ICW1_ICW4 (1 << 0)
#define ICW1_INIT (1 << 4)
#define ICW1_8086 (1 << 0)

#define DISABLE_CODE 0xff

void pic_remap(void)
{
	io_out8(MASTER_COMMAND_SELECTOR, ICW1_INIT | ICW1_ICW4);
	io_wait();
	io_out8(SLAVE_COMMAND_SELECTOR, ICW1_INIT | ICW1_ICW4);
	io_wait();

	io_out8(MASTER_DATA_SELECTOR, MASTER_OFFSET);
	io_wait();
	io_out8(SLAVE_DATA_SELECTOR, SLAVE_OFFSET);
	io_wait();

	io_out8(MASTER_DATA_SELECTOR, MASTER_IDENTITY);
	io_wait();
	io_out8(SLAVE_DATA_SELECTOR, SLAVE_IDENTITY);
	io_wait();

	io_out8(MASTER_DATA_SELECTOR, ICW1_8086);
	io_wait();
	io_out8(SLAVE_DATA_SELECTOR, ICW1_8086);
	io_wait();
}

void pic_disable(void)
{
	io_out8(SLAVE_DATA_SELECTOR, DISABLE_CODE);
	io_wait();
	io_out8(MASTER_DATA_SELECTOR, DISABLE_CODE);
	io_wait();
}

File: hyper_kernel/src/arch/pic.rs
/*
 * Copyright (c) 2022, SkillerRaptor <skillerraptor@protonmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

use log::info;

use crate::arch::io;

const MASTER_COMMAND_SELECTOR: u16 = 0x20;
const MASTER_DATA_SELECTOR: u16 = 0x21;
const MASTER_OFFSET: u8 = 0x20;

const SLAVE_COMMAND_SELECTOR: u16 = 0xA0;
const SLAVE_DATA_SELECTOR: u16 = 0xA1;
const SLAVE_OFFSET: u8 = 0x28;

const ICW1_ICW4: u8 = 1 << 0;
const ICW1_INIT: u8 = 1 << 4;
const ICW1_8086: u8 = 1 << 0;

const MASTER_IDENTITY: u8 = 0x04;
const SLAVE_IDENTITY: u8 = 0x02;

const END_INTERRUPT_CODE: u8 = 0x20;

pub fn remap() {
    io::out8(MASTER_COMMAND_SELECTOR, ICW1_INIT | ICW1_ICW4);
    io::out8(SLAVE_COMMAND_SELECTOR, ICW1_INIT | ICW1_ICW4);

    io::out8(MASTER_DATA_SELECTOR, MASTER_OFFSET);
    io::out8(SLAVE_DATA_SELECTOR, SLAVE_OFFSET);

    io::out8(MASTER_DATA_SELECTOR, MASTER_IDENTITY);
    io::out8(SLAVE_DATA_SELECTOR, SLAVE_IDENTITY);

    io::out8(MASTER_DATA_SELECTOR, ICW1_8086);
    io::out8(SLAVE_DATA_SELECTOR, ICW1_8086);
}

pub fn end_of_interrupt(isr: u8) {
    if isr >= 0x8 {
        io::out8(SLAVE_COMMAND_SELECTOR, END_INTERRUPT_CODE);
    }

    io::out8(MASTER_COMMAND_SELECTOR, END_INTERRUPT_CODE);
}
